#Assignment 3

#Problem 1

\begin{equation}
x_1 + 2x_2 + 3x_3 = 10 \\
5x_2 + 6x_3 = 11 \\
9x_3 = 12
\end{equation}

First we write a function to multiply $N \times N$ matrix $A$ by $N$ element vector $B$. The result will be $N$ element vector $C$, and the general formula for multiplication for i-th element of $C$:

\begin{equation}
C_i = A_{ij} \cdot B_j
\end{equation}

In [ ]:
import numpy as np

In [ ]:
def matrix_mult(A, B): #simple function to multily matrix by vector
  N = B.size
  C = np.zeros(N)
  for i in range(N):
    for j in range(N):
      C[i] += A[i,j]*B[j]
  return C

We can test this for multiplication

$$
\begin{pmatrix}
1 & 2 & 3 \\
4 & 5 & 6 \\
7 & 8 & 9
\end{pmatrix}
\begin{pmatrix}
10 \\
11 \\
12
\end{pmatrix}
$$

In [ ]:
A = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]], dtype=float)
B = np.array([10, 11, 12], dtype=float)
C = matrix_mult(A, B)
print("Resulting matrix = ", C)

Resulting matrix =  [ 68. 167. 266.]


This gives correct result
$
\begin{pmatrix}
68 & 167 & 266
\end{pmatrix}.
$

To solve the above system of equation, we can use backward substitution, as the matrix is upper traigonal. The roots are $N$ elemet matrix $X$. The last root can be found as $X_N = B_N/A_{NN}$. The remaining answers can be found with following formula:

\begin{equation}
X_i = (B_i - \sum_{k=i+1}^{N} A_{ik}X_k)/A_{ii}
\end{equation}

In [ ]:
def back_sub(A, B):
  N = B.size
  X = np.zeros(N)
  X[N-1] = B[N-1]/A[N-1, N-1]
  for i in range(N-2, -1, -1):
    sum = 0
    for k in range(i+1, N):
      sum += A[i,k]*X[k]
    X[i] = (B[i] - sum)/A[i,i]
  return X

In [ ]:
M = np.array([[1, 2, 3], [0, 5, 6], [0, 0, 9]])
b = np.array([10, 11, 12])

ans = back_sub(M, b)
print(ans)

[4.8        0.6        1.33333333]


In the end we get three answers: $x_1 = 4.8$, $x_2=0.6$, $x_3=4/3$.

#Problem 2

$$
\begin{pmatrix}
2.0 & 0.1 & -0.2 \\
0.05 & 4.2 & 0.032 \\
0.12 & -0.07 & 5.0
\end{pmatrix}
\cdot \vec{x}
\begin{pmatrix}
10 \\
11 \\
12
\end{pmatrix}
$$

In order to solve a system of equation $AX=B$ (without pivoting), we must transform matrix $A$ into a triangular matrix with Gauss elimination. First we divide i=1st row of $A$ and i=1st element of $B$ by $A_{ii}. Then for each j-th row and elemnt:

\begin{equation}
A_{jk} - A_{ik} \cdot A_{ji} \\
B_j - B_i \cdot A_{ji}
\end{equation}

In [ ]:
def gauss_elim(A, B):
  N = B.size
  for i in range(N):
    a_ii = A[i, i]
    A[i,:] /= a_ii
    B[i] /= a_ii
    for j in range(i+1, N):
      a_ji = A[j, i]
      A[j, :] -= A[i, :]*a_ji
      B[j] -= B[i]*a_ji
  return A, B

In [ ]:
A2 = np.array([[2.0, 0.1, -0.2], [0.05, 4.2, 0.032], [0.12, -0.07, 5.0]], dtype=float)
B2 = np.array([10, 11, 12], dtype=float)

A2_ans, B2_ans = gauss_elim(A2, B2)
print(A2_ans)
print(B2_ans)

[[ 1.          0.05       -0.1       ]
 [ 0.          1.          0.00881477]
 [ 0.          0.          1.        ]]
[5.         2.56104824 2.31306666]


As we transformed the matrix to upper triangular matrix and can solve the system with backward substitution function from Problem 1.

In [ ]:
X2_ans = back_sub(A2_ans, B2_ans)
print(X2_ans)

[5.10427371 2.54065909 2.31306666]


This gives answers: $x_1 = 5.10427371$, $x_2 = 2.54065909$, $x_3 = 2.31306666$.

#Problem 3

\begin{pmatrix}
6 & 5 & -5 \\
2 & 6 & -2 \\
2 & 5 & -1
\end{pmatrix}

To find eigenvalues and eigenvectors of matrix $A$ with power method, we take an initial guess vector $\textbf{x}$. The eigenvalue can be found after $k$ iterations from ratio $\lambda = r^{(k)} = \frac{A^{k+1}x}{A^k x}$.

In [ ]:
def power_method(A, b, num_iter, error):
  for i in range(num_iter):
    b_new = np.dot(A, b)
    eigenval = np.dot(b_new, b)/np.dot(b, b)
    b = b_new/np.sqrt(b_new[0]**2 + b_new[1]**2 + b_new[2]**2) #for big matrices it is better to calculate norm with more generalized fundtion, not like this

    diff = np.dot(A, b) - eigenval*b
    if np.sqrt(diff[0]**2 + diff[1]**2 + diff[2]**2) < error:
      break
  eigenvec = b
  return eigenval, eigenvec

In [ ]:
A3 = np.array([[6, 5, -5], [2, 6, -2], [2, 5, -1]], dtype=float)
b_test = np.array([-1, 2, 4]) #any initial guess vector

eigenval, eigenvec = power_method(A3, b_test, 100, 1e-8)

print("Largest eigenvalue = ", eigenval)
print("Corresponding eigenvector = ", eigenvec)

Largest eigenvalue =  6.000000011762464
Corresponding eigenvector =  [-0.57735027 -0.57735027 -0.57735027]


Aitken's method takes three approximations of eigenvalues $r_k$, $r_{k+1}$ and $r_{k+2}$, and calculates the eigenvalue as:

\begin{equation}
\lambda = \frac{r_k r_{k+2} - r_{k+1}^2}{r_{k+2} - 2r_{k+1} + r_k}
\end{equation}

In [ ]:
def aitken(eigenvals):
  r_k0 = eigenvals[0]
  r_k1 = eigenvals[1]
  r_k2 = eigenvals[2]

  eigenval = (r_k0*r_k2 - r_k1**2)/(r_k2 - 2*r_k1 + r_k0)

  return eigenval

In [ ]:
eigenvals = []
for i in range(10):
  eigenval2, eigenvec2 = power_method(A3, b_test, 1, 1e-6)
  eigenvals.append(eigenval2)

  if len(eigenvals) >= 3:
    eigenval_aitken = aitken(eigenvals)

print(eigenval_aitken)

nan


<ipython-input-58-e4160fc4e90c>:6: RuntimeWarning: invalid value encountered in scalar divide
  eigenval = (r_k0*r_k2 - r_k1**2)/(r_k2 - 2*r_k1 + r_k0)
